presto-main/src/main/java/com/facebook/presto/execution/resourceGroups/StochasticPriorityQueue.java

```diff
@@ -0,0 +1,261 @@
+        Node<E> candidate = root;
+        while (!candidate.isLeaf()) {
+            long leftTickets = candidate.getLeft().map(Node::getTotalTickets).orElse(0L);
+
+            if (winningTicket < leftTickets) {
+                candidate = candidate.getLeft().get();
+                continue;
+            }
+            else {
+                winningTicket -= leftTickets;
+            }
+
+            if (winningTicket < candidate.getTickets()) {
+                break;
+            }
+            else {
+                winningTicket -= candidate.getTickets();
+            }
+
+            checkState(candidate.getRight().isPresent(), "Expected right node to contain the winner, but it does not exist");
+            candidate = candidate.getRight().get();
+        }
+        checkState(winningTicket < candidate.getTickets(), "Inconsistent winner");
+
+        E value = candidate.getValue();
+        remove(value);
+        return value;
+    }
+
+    @Override
+    public int size()
+    {
+        return index.size();
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return index.isEmpty();
+    }
+
+    private static final class Node<E>
+    {
+        private Optional<Node<E>> parent;
+        private E value;
+        private Optional<Node<E>> left = Optional.empty();
+        private Optional<Node<E>> right = Optional.empty();
+        private int tickets;
+        private long totalTickets;
+        private int decendants;
+
+        private Node(Optional<Node<E>> parent, E value)
+        {
+            this.parent = parent;
+            this.value = value;
+        }
+
+        public E getValue()
+        {
+            return value;
+        }
+
+        public void setValue(E value)
+        {
+            this.value = value;
+        }
+
+        public Optional<Node<E>> getLeft()
+        {
+            return left;
+        }
+
+        public Optional<Node<E>> getRight()
+        {
+            return right;
+        }
+
+        public long getTotalTickets()
+        {
+            return totalTickets;
+        }
+
+        public int getTickets()
+        {
+            return tickets;
+        }
+
+        public void setTickets(int tickets)
+        {
+            checkArgument(tickets > 0, "tickets must be positive");
+            Node<E> node = this;
+            while (node != null) {
+                node.totalTickets -= this.tickets;
+                node.totalTickets += tickets;
+                node = node.parent.orElse(null);
+            }
+            this.tickets = tickets;
+        }
+
+        public boolean isLeaf()
+        {
+            return !left.isPresent() && !right.isPresent();
+        }
+
+        public Node<E> findLeaf()
+        {
+            int leftDecendants = left.map(node -> node.decendants).orElse(0);
+            int rightDecendants = right.map(node -> node.decendants).orElse(0);
+
+            if (leftDecendants > rightDecendants) {
```

Comment:
I'd find this easier if you handled `leftDecendants == 0 && rightDecendants == 0`.. then the rest should be pretty easy since you are assured to have a left or right in the case that one is larger